pyqtorch/noise.py

```diff
@@ -0,0 +1,356 @@
+from pyqtorch.utils import DensityMatrix, density_mat
+
+
+class Noise(torch.nn.Module):
+    def __init__(
+        self, kraus: list[Tensor], target: int, probabilities: tuple[float, ...] | float
+    ) -> None:
+        super().__init__()
+        self.target: int = target
+        self.qubit_support: tuple[int, ...] = (self.target,)
+        self.kraus: list = []
+        for index, tensor in enumerate(kraus):
+            self.register_buffer(f"kraus_{index}", tensor)
+            self.kraus.append(tensor)
+        self._device: torch.device = self.kraus[0].device
+        self.probabilities: tuple[float, ...] | float = probabilities
+
+    @property
+    def name(self) -> str:
+        return self.__class__.__name__
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, type(self)):
+            return self.name == other.name and self.probabilties == other.probabilties
+        return False
+
+    def extra_repr(self) -> str:
+        return f"'qubit_support'={self.qubit_support}, 'probabilities'={self.probabilities}"
+
+    @property
+    def kraus_operator(self) -> list[Tensor]:
+        return self.kraus
+
+    def unitary(self, values: dict[str, Tensor] | Tensor = dict()) -> list[Tensor]:
+        # Since PyQ expects tensor.Size = [2**n_qubits, 2**n_qubits,batch_size].
+        return [kraus_op.unsqueeze(2) for kraus_op in self.kraus]
+
+    def dagger(self, values: dict[str, Tensor] | Tensor = dict()) -> list[Tensor]:
+        return [_dagger(kraus_op) for kraus_op in self.unitary(values)]
+
+    def forward(self, state: Tensor, values: dict[str, Tensor] | Tensor = dict()) -> Tensor:
+        """
+        Applies a noisy quantum channel on the input state.
+        The evolution is represented as a sum of Kraus operators:
+        .. math::
+            S(\\rho) = \\sum_i K_i \\rho K_i^\\dagger,
+
+        Each Kraus operator in the `kraus_list` is applied to the input state, and the result
+        is accumulated to obtain the evolved state.
+
+        Args:
+            state (Tensor): Input quantum state represented as a tensor.
+
+        Returns:
+            Tensor: Quantum state as a density matrix after evolution.
+
+        Raises:
+            TypeError: If the input `state` or `kraus_list` is not a Tensor.
+        """
+        if not isinstance(state, Tensor):
+            raise TypeError("The input must be a Tensor")
+        if not isinstance(state, DensityMatrix):
+            state = density_mat(state)
+        rho_evols: list[Tensor] = []
+        for kraus_unitary, kraus_dagger in zip(self.unitary(values), self.dagger(values)):
+            rho_evol: Tensor = operator_product(
+                kraus_unitary, operator_product(state, kraus_dagger, self.target), self.target
+            )
+            rho_evols.append(rho_evol)
+        rho_final: Tensor = torch.stack(rho_evols, dim=0)
+        rho_final = torch.sum(rho_final, dim=0)
+        return rho_final
```

Comment:
I agree that using a single `einsum` expression could enhance memory efficiency, and I can consider this as a future optimization task.
However, since I also need to work on the implementation of the `pyq` noise in `Qadence`, I believe it would be more practical to address this in a separate pull request focused on refactoring.

In that future pull request, I will consider your idea of using `einsum` and the idea of creating a DensityMatrix class that inherits from `Primitive`.